.goreleaser.yaml

```diff
@@ -29,6 +30,14 @@ checksum:
   name_template: '{{ .ProjectName }}-{{ .Tag }}-checksums.txt'
 snapshot:
   name_template: "{{ .Tag }}-next"
+signs:
+  - cmd: cosign
+    env:
+      - COSIGN_EXPERIMENTAL=1
+    signature: "${artifact}.sig"
+    certificate: "${artifact}.pem"
+    args: ["sign-blob", "--oidc-issuer=https://token.actions.githubusercontent.com", "--output-certificate=${certificate}", "--output-signature=${signature}", "${artifact}"]
```

Comment:
'm trying to understand how signing works here. I don't see any private key being used. What prevents someone from signing a tampered version of the blob with the same command? Is the oidc-issuer taking care of this somehow? (https://token.actions.githubusercontent.com/).

Also, what does the command to verify the blob looks like? Can we document its use in the README somewhere?

Reply:
Yes, this uses keyless signing via OIDC, github provides this OIDC so this is signed by our CI only, cannot be impersonated https://github.com/sigstore/cosign/blob/main/KEYLESS.md#identity-tokens

Manually signing:

Note: This  below opens a browser where you can select your OIDC provider, as this is not automated due to not being on a CI system. On github CI this is automatically done via the github OIDC provider

```
$ COSIGN_EXPERIMENTAL=1 cosign sign-blob --output-certificate=cert.pem --output-signature=sign.sig build/kairos-agent
Using payload from: build/kairos-agent
Generating ephemeral keys...
Retrieving signed certificate...

        Note that there may be personally identifiable information associated with this signed artifact.
        This may include the email address associated with the account with which you authenticate.
        This information will be used for signing this artifact and will be stored in public transparency logs and cannot be removed later.
        By typing 'y', you attest that you grant (or have permission to grant) and agree to have this information stored permanently in transparency logs.

Are you sure you want to continue? (y/[N]): y
Your browser will now be opened to:
https://oauth2.sigstore.dev/auth/auth?access_type=online&client_id=sigstore&code_challenge=SJusQy8JVxZoALD_I0tMWYXEWiEGnHIptmg-sGY9ogU&code_challenge_method=S256&nonce=2KXMHoVrR9E8v9mgx54yzxNvox9&redirect_uri=http%3A%2F%2Flocalhost%3A38763%2Fauth%2Fcallback&response_type=code&scope=openid+email&state=2KXMHmweRBtYeisls9YjBy6HhH2
Successfully verified SCT...
using ephemeral certificate:
-----BEGIN CERTIFICATE-----
MIICpjCCAiugAwIBAgIUZhBhgaGl3aX1I03m6OfZ51TpmrAwCgYIKoZIzj0EAwMw
NzEVMBMGA1UEChMMc2lnc3RvcmUuZGV2MR4wHAYDVQQDExVzaWdzdG9yZS1pbnRl
cm1lZGlhdGUwHhcNMjMwMTE5MDgxOTUzWhcNMjMwMTE5MDgyOTUzWjAAMFkwEwYH
KoZIzj0CAQYIKoZIzj0DAQcDQgAEV2a9NLKJFTpqdSbM6Y/Q8VqIYGhat2CwQ1jT
PIt8rPjYbIiggYP12IvOg+LjfKZKaBDbFnTgtxX6s1kaAheBQ6OCAUowggFGMA4G
A1UdDwEB/wQEAwIHgDATBgNVHSUEDDAKBggrBgEFBQcDAzAdBgNVHQ4EFgQUKE5p
ANR2ebZJwwqSsHLsUi0n8rYwHwYDVR0jBBgwFoAU39Ppz1YkEZb5qNjpKFWixi4Y
ZD8wJQYDVR0RAQH/BBswGYEXaXR4YWthc2VycmFub0BnbWFpbC5jb20wLAYKKwYB
BAGDvzABAQQeaHR0cHM6Ly9naXRodWIuY29tL2xvZ2luL29hdXRoMIGJBgorBgEE
AdZ5AgQCBHsEeQB3AHUA3T0wasbHETJjGR4cmWc3AqJKXrjePK3/h4pygC8p7o4A
AAGFyRzqQQAABAMARjBEAiAVN6lR9MapAXqJwiZ5ocF/zk546CKgnRn6wAGwuAPQ
FAIgaNws113g0b+i+fR7Fgt5tT8tMWdGxox2lzNR8srYGDMwCgYIKoZIzj0EAwMD
aQAwZgIxAIkCfru4lfV5kqa53Fc+fB/AyjE5lWrGX2lzEVaOL+GZCoEemYnhUOIX
TcQ0x9IPXQIxAKUgLoe2pB54QUndvMq+0+HnMiQRKXDBm+IPhmrM3qp1bKXvkGTN
rhkg7CaXDp1s5A==
-----END CERTIFICATE-----

tlog entry created with index: 11499439
Signature wrote in the file sign.sig
Certificate wrote in the file cert.pem
```

Checking the signature:

```
$ COSIGN_EXPERIMENTAL=1 cosign verify-blob build/kairos-agent --signature=sign.sig --cert cert.pem 
tlog entry verified with uuid: b271d6d440a185f8218dd8a22f50b73c14bb7d7d366599f40679e0b62f29c634 index: 11499439
Verified OK
```

In the case of CI builds its even better because you can verify AND check a host of other things, for example:
``
--certificate-github-workflow-name string         contains the workflow claim from the GitHub OIDC Identity token that contains the name of the executed workflow.
      --certificate-github-workflow-ref string          contains the ref claim from the GitHub OIDC Identity token that contains the git ref that the workflow run was based upon.
      --certificate-github-workflow-repository string   contains the repository claim from the GitHub OIDC Identity token that contains the repository that the workflow run was based upon
      --certificate-github-workflow-sha string          contains the sha claim from the GitHub OIDC Identity token that contains the commit SHA that the workflow run was based upon.
      --certificate-github-workflow-trigger string
```

So you can verify the the binary was signed only from the proper repo, sha and workflow

Reply:
Thanks! We should probably document the process of verifying our artifacts for the users.

Reply:
yep, Im struggling where to put it as there is no kairos-agent install page on our docs....maybe I'll create one and put it there.